From git-commits-head-owner@vger.kernel.org Thu Oct 26 20:34:26 2006
Date: Thu, 26 Oct 2006 15:59:09 GMT
Message-Id: <200610261559.k9QFx9tC003499@hera.kernel.org>
From: Linux Kernel Mailing List <linux-kernel@vger.kernel.org>
To: git-commits-head@vger.kernel.org
Subject: [SPARC64]: Fix memory corruption in pci_4u_free_consistent().

commit 012d64ff68f304df1c35ce5902f5023dc14b643f
tree da1e6691c0c9a4d087b441762ce9df8974178339
parent 4130a4b206e7c628482aa12ec30949382c8cdc5e
author David S. Miller <davem@sunset.davemloft.net> 1161840787 -0700
committer David S. Miller <davem@sunset.davemloft.net> 1161841156 -0700

[SPARC64]: Fix memory corruption in pci_4u_free_consistent().

The second argument to free_npages() was being incorrectly
calculated, which would thus access far past the end of the
arena->map[] bitmap.

Signed-off-by: David S. Miller <davem@davemloft.net>

 arch/sparc64/kernel/pci_iommu.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/arch/sparc64/kernel/pci_iommu.c b/arch/sparc64/kernel/pci_iommu.c
index 82e5455..2e7f142 100644
--- a/arch/sparc64/kernel/pci_iommu.c
+++ b/arch/sparc64/kernel/pci_iommu.c
@@ -281,7 +281,7 @@ static void pci_4u_free_consistent(struc
 
 	spin_lock_irqsave(&iommu->lock, flags);
 
-	free_npages(iommu, dvma, npages);
+	free_npages(iommu, dvma - iommu->page_table_map_base, npages);
 
 	spin_unlock_irqrestore(&iommu->lock, flags);
 
